Add Njalla provider

chapatt opened the pull request:
This implements Njalla's new API: https://njal.la/api/

adferrand commented on lexicon/providers/njalla.py:
```diff
@@ -0,0 +1,145 @@
+        return True
+
+    # Helpers
+    def _api_call(self, method, params):
```
Do you have a specific reason to not implement the method `_request` from the base provider? It seems to me that it would fit your case, and you could call `_post` in place of `_api_call`.

chapatt replied:
I expected a comment on this. It just seemed weird to implement the `"request` method when the JSON-RPC API only takes POSTs. If I did, I'd still be tempted to wrap `_post` in another method that takes `method` and `params` args.

I just realized that the API docs are even require login, so some of that might not have been apparent. I'm willing to use `_request` if you feel that's important.

adferrand replied:
I understand that for this provider specifically passing through `_request` may be a burden, but it is for a more general purpose. You see, Lexicon is composed of variety of providers developped by a lot of different people. At some point of Lexicon lifecycle, it is necessary to make some global modifications.

Having some "standard" APIs help a lot to avoid to miss something during these kind of refactoring. In particular, having the actual network requests passing through `_request` is helping to avoid missing specific logic if a global refactoring impact the network operations.

And I am completely ok to wrap the underlying `_request` calls to more suitable high level functions for the Njalla provider.

chapatt replied:
I implemented `_request` :)

adferrand commented on lexicon/providers/njalla.py:
```diff
@@ -0,0 +1,145 @@
+        if len(records) == 1:
+            return records[0]['id']
+
+        raise Exception('Unambiguous record could not be found.')
```
Just to inform you here. You have basically two choices if more than one record is found:
* error out, like you did
* get the first matching record, and warn the user that more than a record was found, and so the first one will be taken

Most of the providers use the second option, but it is your call.

chapatt replied:
I'm using this method in `_delete_record` so deleting the first of multiple matched records seems dangerous.

adferrand replied:
I agree. My option makes more sense in update operation, but for deleting it is bad.

adferrand commented on lexicon/providers/njalla.py:
```diff
@@ -0,0 +1,145 @@
+
+        raise Exception('Unambiguous record could not be found.')
+
+    def _request(self, action='GET', url='/', data=None, query_params=None):
```
See my comment above, about `_api_call`.

adferrand commented on lexicon/tests/providers/test_njalla.py:
```diff
@@ -0,0 +1,22 @@
+        return ['Authorization']
+
+    @pytest.fixture(autouse=True)
+    def _skip_suite(self, request):  # pylint: disable=no-self-use
```
Are you willing to implement the advanced test suite? It is mainly about the capacity to handle multiple records for the same name (like TXTs). It is mostly useful for ACME clients, like Certbot, to issue wildcard certificates, which require two token for the same name.

It is not mandatory, but I wanted to give you an insight about it.

chapatt replied:
I ran into an issue with the very first one: `test_provider_when_calling_create_record_with_duplicate_records_should_be_noop`
- This is not the native behavior of Njalla (it allows duplicate records)
- I attempted doing the check myself with a `_list_records` in the beginning of `_create_record`, but it appeared to not return results for the duplicate record created only a moment before. It seems operations aren't synchronous, and it takes a while for the system to become consistent.

As completely duplicate records are permitted, record sets are also, and are returned correctly by `_list_records` with filter parameters.

I have however not implemented `def test_provider_when_calling_delete_record_with_record_set_name_remove_all`. Is deleting record sets in one go required for by certbot?

adferrand replied:
Alright, may be tricky to match the expected behavior for the advanced test suite. No, deleting all the record set is not mandatory: Certbot will remove a specific TXT+entry+value for each related challenge in a given certificate.

adferrand commented:
Really sorry for the late feedback. Appart `_request`, it is almost done! Don't mind the failure in circle ci. By the way, if you merge master, you will see it disappear.

adferrand commented:
Don't bother the merge from master, I did it for you.

adferrand commented:
Hello @chapatt!

Are you willing to finish the PR? Thanks in advance!

chapatt commented:
Sorry, I lost this PR for a bit. Please let me know if there's anything else that needs fixing up.

chapatt commented:
@adferrand I see I'm failing all of the automated tests now. I can't quite make sense of the errors.

After merging in master, the developer guide local test procedure doesn't seem to work anymore:
```
virtualenv -p python2.7 venv
source venv/bin/activate
py.test lexicon/tests/providers/test_njalla.py
```

adferrand commented:
Hello @chapatt! Really sorry for the delay, I was on vacations. Indeed I changed a little the design of integration tests.

Before you were extending your test class from `IntegrationTest`, with optional explicit markers to disable some parts of the test suite. Now the actual test suite run is defined by the class itself: `IntegrationTestsV2` or `IntegrationTestsV1`. You should use `IntegrationTestsV2` if possible, because the test suite is more complete.

So basically, replace `IntegrationTests` by `IntegrationTestsV2` in your test class, and you are good.

chapatt commented:
> Hello @chapatt! Really sorry for the delay, I was on vacations. Indeed I changed a little the design of integration tests.
> 
> Before you were extending your test class from `IntegrationTest`, with optional explicit markers to disable some parts of the test suite. Now the actual test suite run is defined by the class itself: `IntegrationTestsV2` or `IntegrationTestsV1`. You should use `IntegrationTestsV2` if possible, because the test suite is more complete.
> 
> So basically, replace `IntegrationTests` by `IntegrationTestsV2` in your test class, and you are good.

No worries! I updated to `IntegrationTestsV2`. I still skipped a couple tests because Njalla is not natively capable of them. If they're necessary, it will have to be done client side by querying first, then modifying behavior based on that.

It also looks like I'm failing a couple checks. I can't seem to find useful output of those builds to see why.
